[FrameworkBundle] Fix Routing\DelegatingLoader resiliency to fatal errors

nicolas-grekas opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #14342 |
| License | MIT |
| Doc PR | - |

Tobion commented on src/Symfony/Bundle/FrameworkBundle/Routing/Router.php:
```diff
@@ -51,7 +51,11 @@ public function __construct(ContainerInterface $container, $resource, array $opt
     public function getRouteCollection()
     {
         if (null === $this->collection) {
+            $this->collection = new RouteCollection();
```
This looks like a hack. Is it not possible to not call the method again when it throws an exception?

nicolas-grekas replied:
It is possible, but loading occurs only once. This does not change. What changes is that if the loader fails, we do not retry a second time (which never worked and can't work)

Tobion replied:
It's not the responsibility of the router class to do that. It's the responsibility of the calling code to ensure it does not call `getRouteCollection` twice when the first call throws an exception. This is obvious when you consider that setting an empty RouteCollection does not make sense semantically.

Tobion replied:
Otherwise you could do this logic in basically every method that throws an exception.

nicolas-grekas replied:
The thing is that the getter changes the state of the property. And this state transition is the responsibility of the Router class, not of its caller (which does not know about this).
An other fact is that the loader can fail with a fatal error, which means that the caller will be bypassed.
But the fatal error being caught will be re-injected for display by the Debug component, and this code patch will be taken a second path. Only this code can know about that.

stof replied:
there should at least be a comment explaining the goal of this code

nicolas-grekas replied:
Comment added

Tobion replied:
The getter does not change the state of the property if an exception is thrown. But NOW it does. The next execution will not throw an exception anymore. So this makes it indeterministic from the outside. Again, if you go this its like doing

```
public function foobar() {
   if (0 === rand(0, 1)) { throw new Exception(); }
}
```

fabpot commented on src/Symfony/Component/HttpKernel/EventListener/ExceptionListener.php:
```diff
@@ -65,7 +65,7 @@ public function onKernelException(GetResponseForExceptionEvent $event)
 
         try {
-            $response = $event->getKernel()->handle($request, HttpKernelInterface::SUB_REQUEST, true);
+            $response = $event->getKernel()->handle($request, HttpKernelInterface::SUB_REQUEST, false);
```
This change looks suspicious to me...

nicolas-grekas replied:
Then it should be confirmed. To me, it looks correct: we are dealing with an exception. If an exception is triggered while rendering the page for our original exception, we don't want to trigger a sub-sub-request to generate (recursively) yet an other exception page for the exception page in the exception page :)
We prefer do just catch the second exception and deal with it like. That's the reason why we have a catch block to me.

Tobion replied:
I think this change makes sense. But the `static $handling` variable seems obsolete now. It seems like it was introduced as a workaround for exactly that: https://github.com/nicolas-grekas/symfony/commit/71d5be1d6b239a87837a604b23308bce53fe99ff

nicolas-grekas replied:
I'm going to open an other PR on this topic

Tobion replied:
Please keep PR self-contained. Otherwise it's impossible to see side effects.

nicolas-grekas replied:
See #14369

nicolas-grekas committed:
[FrameworkBundle] Fix Routing\DelegatingLoader resiliency to fatal errors

nicolas-grekas commented:
@Tobion see new patch

fabpot commented:
:+1:

Tobion commented on src/Symfony/Component/Config/Loader/DelegatingLoader.php:
```diff
@@ -38,11 +40,25 @@ public function __construct(LoaderResolverInterface $resolver)
+            // This can happen when a fatal error occurs in ->resolve() or ->load() below
             throw new FileLoaderLoadException($resource);
         }
+        $this->loading[$resource] = true;
```
According to the interface, resource can be `mixed`. So this is not safe.

Tobion commented on src/Symfony/Component/Config/Tests/Loader/DelegatingLoaderTest.php:
```diff
@@ -80,4 +80,21 @@ public function testLoadThrowsAnExceptionIfTheResourceCannotBeLoaded()
+        $resolver = new LoaderResolver(array($loader));
+        $loader = new DelegatingLoader($resolver);
+
+        $loading = new \ReflectionProperty($loader, 'loading');
```
Can we please create a real test that covers functionality instead of implementation?

Tobion commented:
@nicolas-grekas why not do the fix suggested in #14111? Unset https://github.com/nicolas-grekas/symfony/blob/fix-14111/src/Symfony/Component/Config/Loader/FileLoader.php#L104 in the catch block. That makes more sense to me. When an exception in the loading is thrown, the loading state of that resource must be unset.

nicolas-grekas commented:
@Tobion agreed, I updated the FileLoader to release the resource on exceptions.
I also moved the fatal error resiliency to the Routing\DelegatingLoader of the FrameworkBundle.
The very issue with #14342 is failing to load *routes* with double fatal errors (PHP allows recovering from one, but not from two). Thus, patching the DelegatingLoader of the Router component is too generic to me now. Focusing on the DelegatedLoader specifically used for Routing is the right target.

Tobion commented on src/Symfony/Bundle/FrameworkBundle/Routing/DelegatingLoader.php:
```diff
@@ -46,16 +48,23 @@ public function __construct(ControllerNameParser $parser, LoggerInterface $logge
+
+        try {
+            $collection = parent::load($resource, $type);
+            $this->loading = false;
```
this could be after the try...catch to be consistent with the FileLoader

nicolas-grekas replied:
moved after try/catch

Tobion commented on src/Symfony/Bundle/FrameworkBundle/Routing/DelegatingLoader.php:
```diff
@@ -46,16 +48,23 @@ public function __construct(ControllerNameParser $parser, LoggerInterface $logge
     {
-        $collection = parent::load($resource, $type);
+        if (false !== $this->loading) {
+            // This can happen if a fatal error occurs in parent::load()
```
i don't get this. let me recap:
- the first time load is called parent::load throws an exception which is caught and loading is set back to false
- the second time load is called (to render exception page) parent::load throws the same error again of course, but this time the exception is not caught by php??? so loading is NOT set back to false
- the third time load is called (why third time???), this condition will throw an error straight away without calling parent:load

nicolas-grekas replied:
The case handled here is not with an exception but with a fatal error:
- the first time the routes are loaded, a fatal error is triggered (e.g. parse error in controller while loading annotations)
- PHP abruptly empties the stack trace, bypassing all catch blocks. It then calls the registered shutdown functions.
- the ErrorHandler catched the fatal error at this stage and convert it to a FatalErrorException, then re-inject this exception to the HttpKernel with ->terminateWithException (that calls handleException())
- at this stage, if we try to load the routes again, we must prevent the fatal error from occurring a second time, that's why we have this state
- the webprofiler triggers an url generation, which requires the routes being loaded
- this patch handles the case an prevents the second fatal error by triggering an exception beforehand

Tobion replied:
ok can you put that (or more generic) description into the code. otherwise the code is kinda confusing without all that context.

Tobion replied:
But this will now only work with this routing loader. If the user uses a different loader, he has the same problem. So I feel this must be handled generically on top of the loading, e.g. with a decorator.

nicolas-grekas replied:
Comment added

nicolas-grekas replied:
This is specific to the FrameworkBundle because fatal error re-injection is configured in the service definition. If someone has a different loader, the failure will work but the degraded exception page, like today.

nicolas-grekas commented:
The case handled here is not with an exception but with a fatal error:
- the first time the routes are loaded, a fatal error is triggered (e.g. parse error in controller while loading annotations)
- PHP abruptly empties the stack trace, bypassing all catch blocks. It then calls the registered shutdown functions.
- the ErrorHandler catches the fatal error at this stage and converts it to a FatalErrorException, then re-inject this exception to the HttpKernel with ->terminateWithException (that calls handleException())
- at this stage, if we try to load the routes again, we must prevent the fatal error from occurring a second time, that's why we have this state
- the webprofiler triggers an url generation, which requires the routes being loaded
- this patch handles the case and prevents the second fatal error by triggering an exception beforehand

Tobion commented:
Can we not move that logic to the error handler? So when the error handler handles FatalErrorException by reinjecting it to the Kernel::terminateWithException, it will ignore a second fatal error?

nicolas-grekas commented:
We can't: a second fatal error kills the process immediately

Tobion commented:
So what about a decorator implementation? The DelagatingLoader has nothing to do with this new code.
It could be named something like `NoReoccuringFatalErrorLoaderDecorator`

nicolas-grekas commented:
A decorator would create an overhead for everyone, for a case that happen only occasionally while developing, and would very exceptional in prod. This DelagatingLoader already decorated the DelagatingLoader of the router component by inheritance. To me, its fine.

Tobion commented on src/Symfony/Bundle/FrameworkBundle/Routing/DelegatingLoader.php:
```diff
@@ -46,16 +48,39 @@ public function __construct(ControllerNameParser $parser, LoggerInterface $logge
+            //   occurs (e.g. parse error in a controller while loading annotations);
+            // - PHP abruptly empties the stack trace, bypassing all catch blocks;
+            //   it then calls the registered shutdown functions;
+            // - the ErrorHandler catches the fatal error and re-injects its for rendering
```
typo: it

Tobion commented:
:+1:

stof commented on src/Symfony/Bundle/FrameworkBundle/Routing/DelegatingLoader.php:
```diff
@@ -46,16 +47,39 @@ public function __construct(ControllerNameParser $parser, LoggerInterface $logge
     public function load($resource, $type = null)
     {
-        $collection = parent::load($resource, $type);
+        if (false !== $this->loading) {
```
just ``if ($this->loading)``

stof commented on src/Symfony/Bundle/FrameworkBundle/Routing/DelegatingLoader.php:
```diff
@@ -46,16 +47,39 @@ public function __construct(ControllerNameParser $parser, LoggerInterface $logge
+            // This can happen if a fatal error occurs in parent::load().
+            // Here is the scenario:
+            // - while routes are being loaded by parent::load() below, a fatal error
+            //   occurs (e.g. parse error in a controller while loading annotations);
```
what about people not using the DelegatingLoader but a single loader (because they only use a single format for the routes) ?
this can't happen in the fullstack framework unless you redefine the route loader service definitions entirely (it always uses the delegating loader), but it can happen for people using the component standalone

stof replied:
this means it should probably be done in the AnnotationLoader instead

nicolas-grekas replied:
People using the composant standalone will have the fallback exception handler, which works fine.
The pretty-shinny error page is a feature of the framework bundle. Others can take inspiration from it if they want :)

stof commented:
@nicolas-grekas why isn't this logic in the component btw ? It should be avoided when using the standalone component too

nicolas-grekas commented:
If we put the logic in the component, it will break using the DelegatingLoader recursively, if anyone does so.
Moving to AnnotationFileLoader will not fix the resiliency for fatal errors occurring in a ClosureLoader or a PhpFileLoader.

stof commented:
@nicolas-grekas this means that these loaders need the same protection

nicolas-grekas commented:
Or that the delegating loader is the right place to wrap this concern, which is my stand

stof commented:
@nicolas-grekas as I said previously, it can be the right place in the context of the fullstack framework, but not in the standalone usage.

nicolas-grekas commented:
@stof the fatal error resiliency mechanism collides with recursively calling a loader. As such, if I add it to ClosureLoader or PhpFileLoader, it will break anyone's code using them recursively. The FrameworkBundle is responsible for configuring the error handler for its usage. It's no shocking to me that the FrameworkBundle also handles this corollary edge case for fatal errors.
As said, for anyone not using the Routing\DelegatingLoader, the fallback error handling mechanism is perfectly fine. The FrameworkBundle is able to add more DX sugar. That's the only point.

stof commented:
:+1:

WouterJ commented:
I would love to see this included in new bug releases pretty soon. It's been bugging me and many other people as all latest stable versions only show the circular reference exception.
